File: Progs/PhD/src/ILP/examples/src/vb/Facility.vb
' ---------------------------------------------------------------------------
' File: Facility.vb
' Version 12.2  
' ---------------------------------------------------------------------------
' Licensed Materials - Property of IBM
' 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55
' Copyright IBM Corporation 2001, 2010. All Rights Reserved.
' 
' US Government Users Restricted Rights - Use, duplication or
' disclosure restricted by GSA ADP Schedule Contract with
' IBM Corp.
' ---------------------------------------------------------------------------
Imports ILOG.Concert
Imports ILOG.CPLEX

 _


Public Class Facility
   Friend Shared _capacity() As Double
   Friend Shared _fixedCost() As Double
   Friend Shared _cost()() As Double
   
   Friend Shared _nbLocations As Integer
   Friend Shared _nbClients As Integer
   
   
   Friend Shared Sub ReadData(fileName As String)
      System.Console.WriteLine(("Reading data from " + fileName))
      Dim reader As New InputDataReader(fileName)
      
      _capacity = reader.ReadDoubleArray()
      _fixedCost = reader.ReadDoubleArray()
      _cost = reader.ReadDoubleArrayArray()
      
      _nbLocations = _capacity.Length
      _nbClients = _cost.Length
      
      Dim i As Integer
      For i = 0 To _nbClients - 1
         If _cost(i).Length <> _nbLocations Then
         Throw New ILOG.Concert.Exception("inconsistent data in file " + fileName)
         End If
      Next i 
   End Sub 'ReadData
   
   Public Overloads Shared Sub Main(ByVal args() As String)
      Try
         Dim filename As String = "../../../../examples/data/facility.dat"
         If args.Length > 0 Then
            filename = args(0)
         End If
         ReadData(filename)

         Dim cplex As New Cplex()
         Dim open As INumVar() = cplex.BoolVarArray(_nbLocations)

         Dim supply(_nbClients)() As INumVar
         Dim i As Integer
         For i = 0 To _nbClients - 1
            supply(i) = cplex.BoolVarArray(_nbLocations)
         Next i
         For i = 0 To _nbClients - 1
            cplex.AddEq(cplex.Sum(supply(i)), 1)
         Next i
         Dim j As Integer
         For j = 0 To _nbLocations - 1
            Dim v As ILinearNumExpr = cplex.LinearNumExpr()
            For i = 0 To _nbClients - 1
               v.AddTerm(1.0, supply(i)(j))
            Next i
            cplex.AddLe(v, cplex.Prod(_capacity(j), open(j)))
         Next j

         Dim obj As ILinearNumExpr = cplex.ScalProd(_fixedCost, open)
         For i = 0 To _nbClients - 1
            obj.Add(cplex.ScalProd(_cost(i), supply(i)))
         Next i
         cplex.AddMinimize(obj)

         If cplex.Solve() Then
            Dim tolerance As Double = cplex.GetParam(cplex.DoubleParam.EpInt)
            System.Console.WriteLine(("Optimal value: " & cplex.ObjValue))
            For j = 0 To _nbLocations - 1
               If cplex.GetValue(open(j)) >= 1 - tolerance Then
                  System.Console.Write(("Facility " & j & " is open, it serves clients "))
                  For i = 0 To _nbClients - 1
                     If cplex.GetValue(supply(i)(j)) >= 1 - tolerance Then
                        System.Console.Write((" " & i))
                     End If
                  Next i
                  System.Console.WriteLine()
               End If
            Next j
         End If
         cplex.End()
      Catch exc As ILOG.Concert.Exception
         System.Console.WriteLine(("Concert exception '" + exc.ToString + "' caught"))
      Catch exc As System.IO.IOException
         System.Console.WriteLine(("Error reading file " + args(0) + ": " + exc.ToString))
      Catch exc As InputDataReader.InputDataReaderException
         System.Console.WriteLine(exc)
      End Try
   End Sub 'Main
End Class 'Facility

'
'Optimal value: 1383
'Facility 0 is open, it serves clients 2 5 7
'Facility 1 is open, it serves clients 3
'Facility 3 is open, it serves clients 0 1 4 6
'

File: Progs/PhD/src/ILP/examples/src/cs/Facility.cs
// --------------------------------------------------------------------------
// File: Facility.cs
// Version 12.2  
// --------------------------------------------------------------------------
// Licensed Materials - Property of IBM
// 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55
// Copyright IBM Corporation 2003, 2010. All Rights Reserved.
//
// US Government Users Restricted Rights - Use, duplication or
// disclosure restricted by GSA ADP Schedule Contract with
// IBM Corp.
// --------------------------------------------------------------------------

using ILOG.Concert;
using ILOG.CPLEX;


public class Facility {
   internal static double[]   _capacity;
   internal static double[]   _fixedCost;
   internal static double[][] _cost;

   internal static int _nbLocations;
   internal static int _nbClients;

   internal static void ReadData(string fileName) {
      System.Console.WriteLine("Reading data from " + fileName);
      InputDataReader reader = new InputDataReader(fileName);
    
      _capacity  = reader.ReadDoubleArray();
      _fixedCost = reader.ReadDoubleArray();
      _cost      = reader.ReadDoubleArrayArray();
    
      _nbLocations = _capacity.Length;
      _nbClients   = _cost.Length;
    
      for(int i = 0; i < _nbClients; i++)
         if ( _cost[i].Length != _nbLocations )
           throw new ILOG.Concert.Exception("inconsistent data in file " + fileName);
   }

   public static void Main( string[] args ) {
      try {
         string filename  = "../../../../examples/data/facility.dat";
         if (args.Length > 0)
            filename = args[0];
         ReadData(filename);
       
         Cplex cplex = new Cplex();
         INumVar[] open = cplex.BoolVarArray(_nbLocations);
       
         INumVar[][] supply = new INumVar[_nbClients][];
         for(int i = 0; i < _nbClients; i++)
            supply[i] = cplex.BoolVarArray(_nbLocations);
       
         for(int i = 0; i < _nbClients; i++)
            cplex.AddEq(cplex.Sum(supply[i]), 1);
       
         for(int j = 0; j < _nbLocations; j++) {
            ILinearNumExpr v = cplex.LinearNumExpr();
            for(int i = 0; i < _nbClients; i++)
               v.AddTerm(1.0, supply[i][j]);
            cplex.AddLe(v, cplex.Prod(_capacity[j], open[j]));
         }
       
         ILinearNumExpr obj = cplex.ScalProd(_fixedCost, open);
         for(int i = 0; i < _nbClients; i++)
            obj.Add(cplex.ScalProd(_cost[i], supply[i]));
       
         cplex.AddMinimize(obj);
       
         if (cplex.Solve()) {
            double tolerance = cplex.GetParam(Cplex.DoubleParam.EpInt);
            System.Console.WriteLine("Optimal value: " + cplex.ObjValue);
            for(int j = 0; j < _nbLocations; j++) {
               if (cplex.GetValue(open[j]) >= 1 - tolerance) {
                  System.Console.Write("Facility " + j +" is open, it serves clients ");
                  for(int i = 0; i < _nbClients; i++)
                     if (cplex.GetValue(supply[i][j]) >= 1 - tolerance)
                        System.Console.Write(" " + i);
                  System.Console.WriteLine(); 
               }
            }
         }
         cplex.End();
      }
      catch(ILOG.Concert.Exception exc) {
         System.Console.WriteLine("Concert exception '" + exc + "' caught");
      }
      catch (System.IO.IOException exc) {
         System.Console.WriteLine("Error reading file " + args[0] + ": " + exc);
      }
      catch (InputDataReader.InputDataReaderException exc) {
         System.Console.WriteLine(exc);
      }
   }
}

/*
Optimal value: 1383
Facility 0 is open, it serves clients 2 5 7
Facility 1 is open, it serves clients 3
Facility 3 is open, it serves clients 0 1 4 6
*/
